### PR TITLE
Add a method to export membership info to v2 store from RaftCluster

### PR DESCRIPTION
**Pull request by geetasg**
Related to https://github.com/etcd-io/etcd/issues/12913
Introducing method to create the snapshot in v2 format using v3 state. This is used by the backup command in this PR. Main snapshot path will be updated in subsequent PRs.
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: server/etcdserver/api/membership/cluster_test.go
Comment (ahrtr) [inline server/etcdserver/api/membership/cluster_test.go:1003]: Please use go subtest.

**Comment by ahrtr** [comment]
Overall looks good to me with a minor comment.

File: server/etcdserver/api/membership/storev2.go
Comment (serathius) [inline server/etcdserver/api/membership/storev2.go:97]: Don't change this function. The function above `RemoveMember` assumes `isKeyNotFound` causes panic.

Comment (ahrtr) [reply server/etcdserver/api/membership/storev2.go:97]: It isn't a big problem.

- Actually it doesn't matter. Even time when applying conf change, etcd will validate the conf change(search `ValidateConfigurationChange`). It will never run into panic. It's exactly the reason why we never see panic (bug) before
- We will remove the code of adding or removing members into v2store in following PRs soon.

**Comment by ahrtr** [comment]
Please also rebase this PR since it's far behind the head of main.

File: server/etcdserver/api/membership/storev2.go
Comment (ahrtr) [reply server/etcdserver/api/membership/storev2.go:97]: @geetasg Please revert this change, and raise a separate PR, so as not be blocked on this non-important thing.

**Comment by geetasg** [comment]
@ahrtr @serathius thank you for the review. The change to not panic on keyNotFound is primarily driven by the new workflow to export membership to a new store (for snapshot). This store is empty and does not contain the members and will get a keyNotFound if a delete is attempted. The code is now updated to avoid changing mustDeletefromStore method. ptal. Thanks.

**Comment by geetasg** [comment]
cc @ptabor

File: server/etcdserver/api/membership/storev2.go
Comment (ahrtr) [inline server/etcdserver/api/membership/storev2.go:86]: This verification makes sense, but should be in a following PR which removes all the code of adding/removing member in v2store.

File: etcdutl/etcdutl/backup_command.go
Comment (serathius) [inline etcdutl/etcdutl/backup_command.go:178]: You are no longer using `storeData`. Please update the comment and function arguments.

File: server/etcdserver/api/membership/storev2.go
Comment (geetasg) [reply server/etcdserver/api/membership/storev2.go:86]: @ahrtr this is being called only from the RaftCluster Store function which must get an empty store and is only invoked from GetMembershipInfoInV2Format today. This verification is added because the code to assume that member deletion from the store is not needed - it directly adds the removed members to the RemovedMembers prefix in the store. 
I can move the verification to subsequent PR as per review feedback but wanted to clarify that it is protecting the correctness of logic to handle removed members. Please let me know if you would still like me to move this out.

File: etcdutl/etcdutl/backup_command.go
Comment (geetasg) [reply etcdutl/etcdutl/backup_command.go]: updated

File: etcdutl/etcdutl/backup_command.go
Comment (serathius) [inline etcdutl/etcdutl/backup_command.go:170]: Why are we changing `etcdutl backup` command at all? It's a legacy deprecated way to create snapshot.

File: server/etcdserver/api/membership/storev2.go
Comment (ahrtr) [reply server/etcdserver/api/membership/storev2.go:86]: > @ahrtr this is being called only from the RaftCluster Store function which must get an empty store

Yes, I agree. But the `(*RaftCluster) Store` will be called in multiple places in following PR(s). The verification will definitely fail **after** you add more calling of the Store and **before** you remove the code of adding/removing members data in v2store. Currently you only call the `Store` in the etcdutl backup command, so you do not see any verification failure.

A better way should be get it included in a follow PR as the first commit:
- add the verification. We should get a failed verification;
- remove all the code of adding/removing/updating in v2store in the applying code path. Then the verification should be all good.

Since we will resolve the big task "v2store deprecation" in lots of small PRs, let's try to get each PR as simple and standalone as possible.

Comment (geetasg) [reply server/etcdserver/api/membership/storev2.go:86]: >>add the verification. We should get a failed verification;
Adding more clarification - The Store method will not get called directly. It will get called from GetMembershipInfoInV2Format  which should pass empty store as an argument to Store method. The verification should not fail.

File: etcdutl/etcdutl/backup_command.go
Comment (geetasg) [reply etcdutl/etcdutl/backup_command.go:170]: This set of PRs is going towards removing the SetStore method of RaftCluster (along with usage of that store). The backup command is a consumer of SetStore so it needs to be updated. 

If the backup command needs to be removed from code base, I can open another issue to get that done and inform the user base that the backup command is going away in 3.6. Please let me know if the backup command should be removed. /cc @ahrtr

Comment (geetasg) [reply etcdutl/etcdutl/backup_command.go:170]: created:https://github.com/etcd-io/etcd/issues/16321

File: server/etcdserver/api/membership/storev2.go
Comment (ahrtr) [reply server/etcdserver/api/membership/storev2.go:86]: ack. I messed this v2store with [server.go#L252](https://github.com/etcd-io/etcd/blob/9f72c648d3829810f4256ed1868fcf1c7665aa44/server/etcdserver/server.go#L252)

File: etcdutl/etcdutl/backup_command.go
Comment (ahrtr) [reply etcdutl/etcdutl/backup_command.go:170]: > This set of PRs is going towards removing the SetStore method of RaftCluster (along with usage of that store). The backup command is a consumer of SetStore so it needs to be updated.


Agreed. I need to update the `etcdutl backup` command, otherwise we have to keep some source code (e.g. SetStore) which are used by etcdutl only.

> Please let me know if the backup command should be removed. /cc @ahrtr

Let's discuss it separately. I tend to remove it.

File: server/etcdserver/api/membership/cluster_test.go
Comment (ahrtr) [reply server/etcdserver/api/membership/cluster_test.go:1003]: what about this comment?

Comment (geetasg) [reply server/etcdserver/api/membership/cluster_test.go:1003]: updated as per review feedback.

**Review by ahrtr — APPROVED** [review]
LGTM

thx @geetasg

**Comment by geetasg** [comment]
can this PR be merged if there is no additional review feedback? I would like to submit next PR which will consume changes from this one.

**Comment by ahrtr** [comment]
ping @ptabor @serathius @chaochn47 @fuweid

**Review by fuweid — APPROVED** [review]
LGTM

**Comment by ahrtr** [comment]
cc @mitake and @wenjiaswe to take a look as well.

**Comment by ahrtr** [comment]
It's unfortunate that there is no active review on a P0 item for 3.6.

If there is no any additional review feedback by the end of this week. I will merge this PR by then given it's a safe change.

**Comment by wenjiaswe** [comment]
> It's unfortunate that there is no active review on a P0 item for 3.6.

Looks like there has been active review. 

Overall lgtm. @geetasg Thanks for working on this. 

I'll defer to @ahrtr and @serathius to decide on merging.

**Comment by ahrtr** [comment]
thx @wenjiaswe for the review.

**Pull request merged**
